[JIT] Fix isSubtypeOf implementations

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22104 [JIT] Fix DictType isSubtypeOf**

Differential Revision: [D15956726](https://our.internmc.facebook.com/intern/diff/D15956726)

Previously the following would fail:

```
@jit.script
def forward(
    languages: Optional[Dict[int, str]]  
):
    if languages is None:
        print(languages)
```

```
RuntimeError: 
variable 'languages' previously has type Optional[Dict[int, str]] but is now being assigned to a value of type Dict[int, str]:
@jit.script
def forward(
    languages: Optional[Dict[int, str]]  
):
    if languages is None:
       ~~~~~~~~~~~~~~~~~ <--- HERE
        print(languages)
```

Because `Dict[int, str] issubtypeof Optional[Dict[int,str]]` would return false, as we did not fall back to `isSubtypeOf` defined in `Type. This fixes that

Commit:
[JIT] Fix DictType isSubtypeOf

apaszke commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -787,7 +787,7 @@ struct CAFFE2_API DictType : public Type {
       return getKeyType()->isSubtypeOf(dict_rhs->getKeyType()) &&
           getValueType()->isSubtypeOf(dict_rhs->getValueType());
     }
```
We don't need this part now, because `Type::isSubtypeOf` will use `operator==` which contains exactly the same lines.

apaszke commented:
Also, looks like we also don't fall back on `Type::isSubtypeOf` in `OptionalType`.

Commit:
Update on "[JIT] Fix DictType isSubtypeOf"

[JIT] Fix isSubtypeOf implementations

gh-metadata: pytorch pytorch 22104 gh/jamesr66a/11/head

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -265,6 +259,12 @@ struct CAFFE2_API OptionalType: public SingleElementType<TypeKind::OptionalType,
     return create(contained_types[0]);
   }
 
+  bool operator==(const Type& rhs) const override {
```
`Optional[Int] == Option[Number]`?
